File: src/main/java/im/joker/event/ImEventBuilder.java
package im.joker.event;

import im.joker.event.content.state.*;
import im.joker.event.room.AbstractRoomEvent;
import im.joker.event.room.state.*;
import im.joker.helper.IdGenerator;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.Map;
import java.util.UUID;

/**
 * 此类的任何方法都是没有streamId生成
 */
@Component
public class ImEventBuilder {

    @Autowired
    private IdGenerator idGenerator;


    public void setCommonEventFiled(AbstractRoomEvent roomEvent, String roomId, String sender, LocalDateTime time) {
        roomEvent.setEventId(UUID.randomUUID().toString());
        roomEvent.setRoomId(roomId);
        roomEvent.setTransactionId(UUID.randomUUID().toString());
        roomEvent.setSender(sender);
        roomEvent.setOriginServerTs(time);
    }


    /**
     * 创房事件生成
     *
     * @param creator
     * @param roomId
     * @param sender
     * @param time
     * @return
     */
    public RoomCreateEvent roomCreateEvent(String creator, String roomId, String sender, LocalDateTime time) {
        CreateContent createContent = CreateContent.builder()
                .creator(creator)
                .mFederate(false)
                .roomVersion(idGenerator.roomVersion())
                .build();
        RoomCreateEvent roomCreateEvent = RoomCreateEvent.builder()
                .content(createContent)
                .type(EventType.Creation.getId())
                .stateKey("")
                .build();
        setCommonEventFiled(roomCreateEvent, roomId, sender, time);
        return roomCreateEvent;
    }

    /**
     * 成员变动事件生成
     *
     * @param roomId
     * @param time
     * @param sender
     * @param reason
     * @param stateKey
     * @param displayName
     * @param avatarUrl
     * @param membership
     * @return
     */
    public MembershipEvent membershipEvent(String roomId, LocalDateTime time, String sender, String reason,
                                           String stateKey, String displayName, String avatarUrl, MembershipType membership) {
        MembershipContent mContent = MembershipContent
                .builder()
                .avatarUrl(avatarUrl)
                .reason(reason)
                .membership(membership.name().toLowerCase())
                .displayName(displayName)
                .build();

        MembershipEvent membershipEvent = MembershipEvent.builder()
                .type(EventType.Membership.getId())
                .stateKey(stateKey)
                .content(mContent).build();
        setCommonEventFiled(membershipEvent, roomId, sender, time);
        return membershipEvent;
    }

    /**
     * 房间命名事件生成
     *
     * @param roomName
     * @param roomId
     * @param sender
     * @param time
     * @return
     */
    public RoomNameEvent roomNameEvent(String roomName, String roomId, String sender, LocalDateTime time) {
        RoomNameContent roomNameContent = RoomNameContent.builder()
                .name(roomName).build();
        RoomNameEvent roomNameEvent = RoomNameEvent.builder()
                .content(roomNameContent)
                .type(EventType.RoomName.getId())
                .stateKey("")
                .build();
        setCommonEventFiled(roomNameEvent, roomId, sender, time);
        return roomNameEvent;
    }

    public RoomTopicEvent roomTopicEvent(String topic, String roomId, String sender, LocalDateTime time) {
        RoomTopicContent topicContent = RoomTopicContent.builder().topic(topic).build();
        RoomTopicEvent topicEvent = RoomTopicEvent.builder().content(topicContent)
                .type(EventType.RoomTopic.getId())
                .stateKey("").build();
        setCommonEventFiled(topicEvent, roomId, sender, time);
        return topicEvent;
    }

    /**
     * 房间权限定义事件生成
     *
     * @param ban
     * @param invite
     * @param kick
     * @param redact
     * @param state
     * @param events
     * @param message
     * @param room
     * @param users
     * @param roomId
     * @param sender
     * @param time
     * @return
     */
    public PowerLevelEvent powerLevelEvent(int ban, int invite, int kick, int redact, int state, Map<String, Integer> events, int userDefault,
                                           int message, int room, Map<String, Integer> users, String roomId, String sender,
                                           LocalDateTime time) {
        PowerLevelContent pc = PowerLevelContent.builder()
                .ban(ban)
                .invite(invite)
                .kick(kick)
                .events(events)
                .redact(redact)
                .stateDefault(state)
                .eventDefault(message)
                .users(users)
                .userDefault(userDefault)
                .notifications(PowerLevelContent.RoomNotificationPower.builder().room(room).build())
                .build();
        PowerLevelEvent pe = PowerLevelEvent.builder()
                .type(EventType.PowerLevel.getId())
                .content(pc).build();
        setCommonEventFiled(pe, roomId, sender, time);
        return pe;
    }

    /**
     * 创房时默认发的powerLevelEvent
     *
     * @param roomId
     * @param creator
     * @param time
     * @return
     */
    public PowerLevelEvent defaultPowerLevelEvent(String roomId, String creator, LocalDateTime time) {
        Map<String, Integer> users = Map.of(creator, 100);
        return powerLevelEvent(60, 50, 60, 50, 50, null, 50, 0, 50, users, roomId, creator, time);
    }

    public RoomJoinRuleEvent roomJoinRuleEvent(RoomJoinRuleType roomJoinRuleType, String roomId, String sender, LocalDateTime time) {
        RoomJoinRuleContent content = RoomJoinRuleContent
                .builder()
                .joinRule(roomJoinRuleType.name().toLowerCase())
                .build();

        RoomJoinRuleEvent event = RoomJoinRuleEvent.builder()
                .type(EventType.RoomJoinRule.getId())
                .content(content)
                .build();
        setCommonEventFiled(event, roomId, sender, time);
        return event;
    }

    /**
     * 创房时默认的房间历史是否可以查看事件
     *
     * @param roomId
     * @param sender
     * @return
     */
    public HistoryVisibilityEvent defaultHistoryVisibilityEvent(String roomId, String sender, LocalDateTime time) {
        HistoryVisibilityContent hvc = HistoryVisibilityContent.builder().historyVisibility("joined").build();
        HistoryVisibilityEvent hvEvent = HistoryVisibilityEvent.builder().content(hvc).build();
        setCommonEventFiled(hvEvent, roomId, sender, time);
        return hvEvent;
    }
}

File: src/main/kotlin/im/joker/helper/ImEventBuilder.kt
package im.joker.helper

import im.joker.api.vo.room.TypingRequest
import im.joker.device.Device
import im.joker.event.EventType
import im.joker.event.MembershipType
import im.joker.event.RoomJoinRuleType
import im.joker.event.content.other.FullReadContent
import im.joker.event.content.other.ReceiptContent
import im.joker.event.content.other.TypingContent
import im.joker.event.content.state.*
import im.joker.event.room.other.ReceiptEvent
import im.joker.event.room.other.TypingEvent
import im.joker.event.room.state.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.time.LocalDateTime
import java.util.*

@Component
class ImEventBuilder {

    @Autowired
    lateinit var idGenerator: IdGenerator


    suspend fun createRoomEvent(creator: String, roomId: String, createTime: LocalDateTime): RoomCreateEvent {

        val createContent = CreateContent().apply {
            this.creator = creator
            roomVersion = idGenerator.roomVersion()
            mFederate = false
        }
        return RoomCreateEvent().apply {
            content = createContent
            sender = creator
            streamId = idGenerator.nextEventStreamId()
            stateKey = ""
            type = EventType.Creation.id
            transactionId = UUID.randomUUID().toString()
            originServerTs = createTime
            this.roomId = roomId
            eventId = UUID.randomUUID().toString()
        }

    }

    suspend fun membershipEvent(roomId: String, time: LocalDateTime, sender: String,
                                reason: String, stateKey: String, displayName: String,
                                avatarUrl: String, membership: MembershipType): MembershipEvent {
        val membershipContent = MembershipContent().apply {
            this.avatarUrl = avatarUrl
            this.reason = reason
            this.membership = membership.id
            this.displayName = displayName
        }
        return MembershipEvent().apply {
            this.type = EventType.Membership.id
            this.stateKey = stateKey
            content = membershipContent
            this.sender = sender
            streamId = idGenerator.nextEventStreamId()
            transactionId = UUID.randomUUID().toString()
            originServerTs = time
            this.roomId = roomId
            eventId = UUID.randomUUID().toString()
        }

    }

    suspend fun powerDefEvent(roomId: String, creator: String, now: LocalDateTime): PowerLevelEvent {
        val users = mapOf(creator to 100)
        return powerLevelEvent(60, 50, 60, 50, 50, null,
                50, 0, 50, users, roomId, creator, now)
    }

    /**
     * 房间权限定义事件生成
     *
     * @param ban
     * @param invite
     * @param kick
     * @param redact
     * @param state
     * @param events
     * @param message
     * @param room
     * @param users
     * @param roomId
     * @param sender
     * @param time
     * @return
     */
    suspend fun powerLevelEvent(ban: Int, invite: Int, kick: Int, redact: Int, state: Int, events: Map<String, Int>?, userDefault: Int,
                                message: Int, room: Int, users: Map<String, Int>, roomId: String?, sender: String,
                                time: LocalDateTime): PowerLevelEvent {
        val pc: PowerLevelContent = PowerLevelContent().apply {
            this.ban = ban
            this.invite = invite
            this.kick = kick
            this.events = events
            this.redact = redact
            this.stateDefault = state
            this.eventDefault = message
            this.users = users
            this.userDefault = userDefault
            this.notifications = PowerLevelContent.RoomNotificationPower().apply { this.room = room }
        }

        return PowerLevelEvent().apply {
            this.type = EventType.PowerLevel.id
            this.stateKey = ""
            this.content = pc
            this.sender = sender
            streamId = idGenerator.nextEventStreamId()
            transactionId = UUID.randomUUID().toString()
            originServerTs = time
            this.roomId = roomId
            eventId = UUID.randomUUID().toString()
        }

    }

    /**
     * 创房时默认的房间历史是否可以查看事件
     *
     * @param roomId
     * @param sender
     * @return
     */
    suspend fun defaultHistoryVisibilityEvent(roomId: String?, sender: String?, time: LocalDateTime): HistoryVisibilityEvent {
        val hvc: HistoryVisibilityContent = HistoryVisibilityContent().apply {
            historyVisibility = "joined"
        }
        return HistoryVisibilityEvent().apply {
            stateKey = ""
            type = EventType.HistoryVisibility.id
            content = hvc
            this.sender = sender
            streamId = idGenerator.nextEventStreamId()
            transactionId = UUID.randomUUID().toString()
            originServerTs = time
            this.roomId = roomId
            eventId = UUID.randomUUID().toString()
        }
    }

    suspend fun roomJoinRuleEvent(joinRuleType: RoomJoinRuleType, roomId: String, sender: String, time: LocalDateTime): RoomJoinRuleEvent {
        val content = RoomJoinRuleContent().apply {
            this.joinRule = joinRuleType.id
        }
        return RoomJoinRuleEvent().apply {
            stateKey = ""
            type = EventType.RoomJoinRule.id
            this.content = content
            this.sender = sender
            streamId = idGenerator.nextEventStreamId()
            transactionId = UUID.randomUUID().toString()
            originServerTs = time
            this.roomId = roomId
            eventId = UUID.randomUUID().toString()
        }
    }

    suspend fun roomNameEvent(roomName: String, roomId: String, sender: String, time: LocalDateTime): RoomNameEvent {
        val roomNameContent: RoomNameContent = RoomNameContent().apply {
            name = roomName
        }
        return RoomNameEvent().apply {
            stateKey = ""
            type = EventType.RoomName.id
            content = roomNameContent
            this.sender = sender
            streamId = idGenerator.nextEventStreamId()
            transactionId = UUID.randomUUID().toString()
            originServerTs = time
            this.roomId = roomId
            eventId = UUID.randomUUID().toString()
        }
    }

    suspend fun roomTopicEvent(topic: String, roomId: String, sender: String, time: LocalDateTime): RoomTopicEvent {
        val topicContent: RoomTopicContent = RoomTopicContent().apply {
            this.topic = topic
        }
        return RoomTopicEvent().apply {
            content = topicContent
            stateKey = ""
            type = EventType.RoomTopic.id
            this.sender = sender
            streamId = idGenerator.nextEventStreamId()
            transactionId = UUID.randomUUID().toString()
            originServerTs = time
            this.roomId = roomId
            eventId = UUID.randomUUID().toString()
        }
    }


    suspend fun receiptEvent(roomId: String, read: String, now: LocalDateTime, loginDevice: Device): ReceiptEvent {
        val eventId = UUID.randomUUID().toString()
        val content = ReceiptContent().apply {
            properties[eventId] = ReceiptContent.Receipts().apply {
                val receipt = ReceiptContent.Receipt().apply {
                    this.ts = now
                }
                this.read[loginDevice.userId] = receipt
            }
        }
        return ReceiptEvent().apply {
            this.content = content
            this.roomId = roomId
            type = EventType.Receipt.id
            this.sender = loginDevice.userId
            streamId = idGenerator.nextEventStreamId()
            transactionId = UUID.randomUUID().toString()
            originServerTs = now
            this.eventId = UUID.randomUUID().toString()

        }
    }

    suspend fun typingEvent(sender: String, typingRequest: TypingRequest, time: LocalDateTime): TypingEvent {
        val typingContent = TypingContent().apply {
            this.timeout = typingRequest.timeout
            this.typing = typingRequest.typing
        }
        return TypingEvent().apply {
            this.content = typingContent
            this.roomId = typingRequest.roomId
            type = EventType.Typing.id
            this.sender = sender
            originServerTs = time
        }
    }

    fun roomAvatarEvent(): RoomAvatarEvent? {
        return null
    }

}